File: app/src/main/java/com/android/benchmark/ui/TextScrollActivity.java
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an AS IS BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.benchmark.ui;

import android.content.Intent;
import android.os.Bundle;
import android.widget.ArrayAdapter;
import android.widget.ListAdapter;
import android.widget.ListView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.android.benchmark.registry.BenchmarkRegistry;
import com.android.benchmark.ui.automation.Automator;
import com.android.benchmark.ui.automation.Interaction;

import java.io.File;

public class TextScrollActivity extends ListActivityBase {

    public static final String EXTRA_HIT_RATE = ".TextScrollActivity.EXTRA_HIT_RATE";

    private static final int PARAGRAPH_COUNT = 200;

    private int mHitPercentage = 100;
    @Nullable
    private Automator mAutomator;
    private String mName;

    @Override
    public void onCreate(final Bundle savedInstanceState) {
        this.mHitPercentage = this.getIntent().getIntExtra(TextScrollActivity.EXTRA_HIT_RATE,
                this.mHitPercentage);
        super.onCreate(savedInstanceState);
        int runId = this.getIntent().getIntExtra("com.android.benchmark.RUN_ID", 0);
        int iteration = this.getIntent().getIntExtra("com.android.benchmark.ITERATION", -1);
        int id = this.getIntent().getIntExtra(BenchmarkRegistry.EXTRA_ID, -1);

        if (-1 == id) {
            this.finish();
            return;
        }

        this.mName = BenchmarkRegistry.getBenchmarkName(this, id);

        this.mAutomator = new Automator(mName, runId, iteration, this.getWindow(),
                new Automator.AutomateCallback() {
            @Override
            public void onPostAutomate() {
                final Intent result = new Intent();
                TextScrollActivity.this.setResult(Activity.RESULT_OK, result);
                TextScrollActivity.this.finish();
            }

            @Override
            public void onAutomate() {
                final ListView v = TextScrollActivity.this.findViewById(android.R.id.list);

                final int[] coordinates = new int[2];
                v.getLocationOnScreen(coordinates);

                final int x = coordinates[0];
                final int y = coordinates[1];

                final float width = v.getWidth();
                final float height = v.getHeight();

                final float middleX = (x + width) / 2;
                final float middleY = (y + height) / 2;

                final Interaction flingUp = Interaction.newFlingUp(middleX, middleY);
                final Interaction flingDown = Interaction.newFlingDown(middleX, middleY);

                this.addInteraction(flingUp);
                this.addInteraction(flingDown);
                this.addInteraction(flingUp);
                this.addInteraction(flingDown);
                this.addInteraction(flingUp);
                this.addInteraction(flingDown);
                this.addInteraction(flingUp);
                this.addInteraction(flingDown);
            }
        });

        this.mAutomator.start();
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (null != mAutomator) {
            this.mAutomator.cancel();
            this.mAutomator = null;
        }
    }

    @NonNull
    @Override
    protected ListAdapter createListAdapter() {
        return new ArrayAdapter<>(this, android.R.layout.simple_list_item_1,
                Utils.buildParagraphListWithHitPercentage(TextScrollActivity.PARAGRAPH_COUNT, 80));
    }

    @Override
    protected String getName() {
        return this.mName;
    }
}

File: app/src/main/java/com/android/benchmark/ui/TextScrollActivity.kt
/*
 * Copyright (C) 2015 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the License);
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an AS IS BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.benchmark.ui

android.os.Bundleimport android.widget.ArrayAdapterimport android.widget.ListAdapterimport android.widget.ListViewimport com.android.benchmark.registry.BenchmarkRegistryimport com.android.benchmark.ui.automation.Automatorimport com.android.benchmark.ui.automation.Automator.AutomateCallbackimport com.android.benchmark.ui.automation.Interaction
class TextScrollActivity : ListActivityBase() {
    private var mHitPercentage = 100
    private var mAutomator: Automator? = null
    override var name: String? = null
        private set

    override fun onCreate(savedInstanceState: Bundle?) {
        mHitPercentage = intent.getIntExtra(EXTRA_HIT_RATE,
                mHitPercentage)
        super.onCreate(savedInstanceState)
        val runId = intent.getIntExtra("com.android.benchmark.RUN_ID", 0)
        val iteration = intent.getIntExtra("com.android.benchmark.ITERATION", -1)
        val id = intent.getIntExtra(BenchmarkRegistry.Companion.EXTRA_ID, -1)
        if (-1 == id) {
            finish()
            return
        }
        name = BenchmarkRegistry.Companion.getBenchmarkName(this, id)
        mAutomator = Automator(name, runId, iteration, window,
                object : AutomateCallback() {
                    override fun onPostAutomate() {
                        val result = Intent()
                        setResult(RESULT_OK, result)
                        finish()
                    }

                    override fun onAutomate() {
                        val v = findViewById<ListView>(android.R.id.list)
                        val coordinates = IntArray(2)
                        v.getLocationOnScreen(coordinates)
                        val x = coordinates[0]
                        val y = coordinates[1]
                        val width = v.width.toFloat()
                        val height = v.height.toFloat()
                        val middleX = (x + width) / 2
                        val middleY = (y + height) / 2
                        val flingUp: Interaction = Interaction.Companion.newFlingUp(middleX, middleY)
                        val flingDown: Interaction = Interaction.Companion.newFlingDown(middleX, middleY)
                        addInteraction(flingUp)
                        addInteraction(flingDown)
                        addInteraction(flingUp)
                        addInteraction(flingDown)
                        addInteraction(flingUp)
                        addInteraction(flingDown)
                        addInteraction(flingUp)
                        addInteraction(flingDown)
                    }
                })
        mAutomator!!.start()
    }

    override fun onPause() {
        super.onPause()
        if (null != mAutomator) {
            mAutomator!!.cancel()
            mAutomator = null
        }
    }

    override fun createListAdapter(): ListAdapter {
        return ArrayAdapter<String?>(this, android.R.layout.simple_list_item_1,
                Utils.Companion.buildParagraphListWithHitPercentage(PARAGRAPH_COUNT, 80))
    }

    companion object {
        const val EXTRA_HIT_RATE = ".TextScrollActivity.EXTRA_HIT_RATE"
        private const val PARAGRAPH_COUNT = 200
    }
}